File: shared/bindings/textcore.py
"""Python backend binding for the shared WASM core (textcore.wasm).

This is the SINGLE integration point the backend uses to call shared algorithms.
It loads the same .wasm the browser loads, so any operator added here is
automatically available — and identical — on both ends.

Requires: wasmtime  (pip install wasmtime)
"""
import os
import numpy as np
from wasmtime import Store, Module, Instance, Memory

_REPO = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
_WASM = os.path.join(_REPO, "shared", "build", "textcore.wasm")


class TextCore:
    def __init__(self, wasm_path: str = _WASM):
        self.store = Store()
        self.module = Module.from_file(self.store.engine, wasm_path)
        self.inst = Instance(self.store, self.module, [])
        self.ex = self.inst.exports(self.store)
        self.mem = self.ex["memory"]

    def _alloc(self, nbytes: int) -> int:
        return int(self.ex["alloc"](self.store, nbytes))

    def _free(self, ptr: int, nbytes: int) -> None:
        self.ex["dealloc"](self.store, ptr, nbytes)

    def distance_transform_edt(self, mask_u8: np.ndarray, h: int, w: int) -> np.ndarray:
        """Distance to nearest TEXT (nonzero) pixel. Mirrors
        cv2.distanceTransform((cur == 0), DIST_L2, 3) semantics but exact."""
        n = h * w
        p_mask = self._alloc(n)
        p_out = self._alloc(n * 4)
        try:
            self.mem.write(self.store, np.ascontiguousarray(mask_u8, dtype=np.uint8).tobytes(), p_mask)
            self.ex["distance_transform_edt"](self.store, p_mask, h, w, p_out)
            out_bytes = bytes(self.mem.read(self.store, p_out, p_out + n * 4))
        finally:
            self._free(p_mask, n)
            self._free(p_out, n * 4)
        return np.frombuffer(out_bytes, dtype=np.float32).reshape(h, w).copy()

    def rgb_to_gray(self, rgb, h: int, w: int) -> np.ndarray:
        """RGB (H*W*3, any dtype 0..255) -> grayscale u8 (H*W)."""
        n = h * w
        arr = np.ascontiguousarray(rgb, dtype=np.float32)
        p_in = self._alloc(n * 3 * 4)
        p_out = self._alloc(n)
        try:
            self.mem.write(self.store, arr.tobytes(), p_in)
            self.ex["rgb_to_gray"](self.store, p_in, h, w, p_out)
            out = np.frombuffer(bytes(self.mem.read(self.store, p_out, p_out + n)), dtype=np.uint8).reshape(h, w).copy()
        finally:
            self._free(p_in, n * 3 * 4)
            self._free(p_out, n)
        return out

    def threshold_otsu(self, u8, h: int, w: int):
        """Otsu threshold. Returns (thr: float, bin: u8 H*W with 0/255)."""
        n = h * w
        p_in = self._alloc(n)
        p_out = self._alloc(n)
        try:
            self.mem.write(self.store, np.ascontiguousarray(u8, dtype=np.uint8).tobytes(), p_in)
            thr = float(self.ex["threshold_otsu"](self.store, p_in, p_out, n))
            bin = np.frombuffer(bytes(self.mem.read(self.store, p_out, p_out + n)), dtype=np.uint8).reshape(h, w).copy()
        finally:
            self._free(p_in, n)
            self._free(p_out, n)
        return thr, bin

    def morphology(self, mask_u8, h: int, w: int, kern_u8, kh: int, kw: int, op: str = "dilate") -> np.ndarray:
        """Binary morphology on a 0/1 (or 0/255) mask using an explicit kernel bitmap.

        `kern_u8` is a kh*kw u8 array (1 = include / 0 = skip) — e.g. the exact
        cv2.getStructuringElement output. Returns 0/1 u8 H*W. op: "dilate"/"erode".
        """
        n = h * w
        nk = kh * kw
        op_code = 0 if op == "erode" else 1
        p_in = self._alloc(n)
        p_out = self._alloc(n)
        p_k = self._alloc(nk)
        try:
            self.mem.write(self.store, np.ascontiguousarray(mask_u8, dtype=np.uint8).tobytes(), p_in)
            self.mem.write(self.store, np.ascontiguousarray(kern_u8, dtype=np.uint8).tobytes(), p_k)
            self.ex["morphology"](self.store, p_in, p_out, h, w, p_k, kh, kw, op_code)
            out = np.frombuffer(bytes(self.mem.read(self.store, p_out, p_out + n)), dtype=np.uint8).reshape(h, w).copy()
        finally:
            self._free(p_in, n)
            self._free(p_out, n)
            self._free(p_k, nk)
        return out

    def connected_components(self, mask_u8, h: int, w: int):
        """8-connected components. Returns (n, labels int32 H*W, stats list of dicts).
        stats[0] is the background placeholder."""
        n = h * w
        p_in = self._alloc(n)
        p_labels = self._alloc(n * 4)
        try:
            self.mem.write(self.store, np.ascontiguousarray(mask_u8, dtype=np.uint8).tobytes(), p_in)
            ncomp = int(self.ex["connected_components"](self.store, p_in, p_labels, h, w))
            labels = np.frombuffer(bytes(self.mem.read(self.store, p_labels, p_labels + n * 4)), dtype=np.int32).reshape(h, w).copy()
        finally:
            self._free(p_in, n)
            # keep p_labels for the stats call below
        p_stats = self._alloc(ncomp * 5 * 4)
        try:
            self.ex["connected_components_stats"](self.store, p_labels, p_stats, h, w, ncomp)
            raw = np.frombuffer(bytes(self.mem.read(self.store, p_stats, p_stats + ncomp * 5 * 4)), dtype=np.int32).reshape(ncomp, 5).copy()
        finally:
            self._free(p_stats, ncomp * 5 * 4)
            self._free(p_labels, n * 4)
        stats = [{"left": int(r[0]), "top": int(r[1]), "width": int(r[2]),
                  "height": int(r[3]), "area": int(r[4])} for r in raw]
        return ncomp, labels, stats

    def resize_gray_cubic(self, u8, h2: int, w2: int) -> np.ndarray:
        h, w = u8.shape[:2]
        n = h * w
        n2 = h2 * w2
        p_in = self._alloc(n)
        p_out = self._alloc(n2)
        try:
            self.mem.write(self.store, np.ascontiguousarray(u8, dtype=np.uint8).tobytes(), p_in)
            self.ex["resize_gray_cubic"](self.store, p_in, p_out, h, w, h2, w2)
            out = np.frombuffer(bytes(self.mem.read(self.store, p_out, p_out + n2)), dtype=np.uint8).reshape(h2, w2).copy()
        finally:
            self._free(p_in, n)
            self._free(p_out, n2)
        return out

    def resize_float_linear(self, f32, h2: int, w2: int) -> np.ndarray:
        h, w = f32.shape[:2]
        n = h * w
        n2 = h2 * w2
        p_in = self._alloc(n * 4)
        p_out = self._alloc(n2 * 4)
        try:
            self.mem.write(self.store, np.ascontiguousarray(f32, dtype=np.float32).tobytes(), p_in)
            self.ex["resize_float_linear"](self.store, p_in, p_out, h, w, h2, w2)
            out = np.frombuffer(bytes(self.mem.read(self.store, p_out, p_out + n2 * 4)), dtype=np.float32).reshape(h2, w2).copy()
        finally:
            self._free(p_in, n * 4)
            self._free(p_out, n2 * 4)
        return out

    def patchmatch_inpaint(self, rgb_f32, h: int, w: int, mask_u8,
                           sample_u8=None, p: int = 7, direction_deg: float = -1.0,
                           seed: int = 0) -> np.ndarray:
        """PatchMatch inpainting — THE shared fill routine (browser + backend call this).

        rgb_f32 : H*W*3 float32 (values 0..255). Mutated to the filled result.
        mask_u8 : H*W, >0 = hole to fill.
        sample_u8 : optional H*W, >0 = allowed source region.
        p : patch size (odd). direction_deg : < -1e30 disables direction mode.
        Returns H*W*3 float32 filled image.
        """
        n = h * w
        arr = np.ascontiguousarray(rgb_f32, dtype=np.float32)
        m = np.ascontiguousarray(mask_u8, dtype=np.uint8)
        has_sample = 0
        p_sample = 0
        p_s = 0
        if sample_u8 is not None:
            has_sample = 1
            s = np.ascontiguousarray(sample_u8, dtype=np.uint8)
            p_s = self._alloc(n)
            self.mem.write(self.store, s.tobytes(), p_s)
        p_in = self._alloc(n * 3 * 4)
        p_out = self._alloc(n * 3 * 4)
        p_mask = self._alloc(n)
        try:
            self.mem.write(self.store, arr.tobytes(), p_in)
            self.mem.write(self.store, m.tobytes(), p_mask)
            self.ex["patchmatch_inpaint"](
                self.store, p_in, h, w, p_mask, p_s, has_sample, p,
                float(direction_deg), int(seed) & 0xFFFFFFFF, p_out)
            out = np.frombuffer(bytes(self.mem.read(self.store, p_out, p_out + n * 3 * 4)),
                               dtype=np.float32).reshape(h, w, 3).copy()
        finally:
            self._free(p_in, n * 3 * 4)
            self._free(p_out, n * 3 * 4)
            self._free(p_mask, n)
            if p_s:
                self._free(p_s, n)
        return out

    def synthesize_masks(self, text_mask_u8, h: int, w: int, edge: int = 1,
                        limit_u8=None) -> tuple:
        """Text-mask synthesis — THE shared operator (browser + backend call this).

        text_mask_u8 : H*W, >0 = text. edge : >0 dilate / <0 erode / 0 identity
                      (diameter = abs(edge)*2+1 ellipse). limit_u8 : optional H*W,
                      >0 restricts the fill region (sample unaffected).
        Returns (fill_mask H*W u8 0/255, sample_mask H*W u8 0/255) where
        sample = whole - fill.
        """
        n = h * w
        tm = np.ascontiguousarray(text_mask_u8, dtype=np.uint8)
        has_limit = 0
        p_lim = 0
        if limit_u8 is not None:
            has_limit = 1
            p_lim = self._alloc(n)
            self.mem.write(self.store, np.ascontiguousarray(limit_u8, dtype=np.uint8).tobytes(), p_lim)
        p_in = self._alloc(n)
        p_fill = self._alloc(n)
        p_smpl = self._alloc(n)
        try:
            self.mem.write(self.store, tm.tobytes(), p_in)
            self.ex["synthesize_masks"](
                self.store, p_in, h, w, edge, p_lim, has_limit, p_fill, p_smpl)
            fill = np.frombuffer(bytes(self.mem.read(self.store, p_fill, p_fill + n)),
                                dtype=np.uint8).reshape(h, w).copy()
            sample = np.frombuffer(bytes(self.mem.read(self.store, p_smpl, p_smpl + n)),
                                  dtype=np.uint8).reshape(h, w).copy()
        finally:
            self._free(p_in, n)
            self._free(p_fill, n)
            self._free(p_smpl, n)
            if p_lim:
                self._free(p_lim, n)
        return fill, sample

    def deglow_full_green_v2(self, rgb_f32, h: int, w: int, tmask_u8,
                            strength: float = 1.0, zone_ratio: float = 0.6,
                            zone_expand: int = 0, protect_px: int = 0,
                            chroma_keep: int = 0) -> tuple:
        """De-glow (full green v2) — THE shared operator (browser + backend call this).

        rgb_f32  : H*W*3 float32 (0..255).
        tmask_u8 : H*W, >0 = known text.
        Returns (clean H*W*3 u8, core_mask H*W u8 0/255).
        """
        n = h * w
        arr = np.ascontiguousarray(rgb_f32, dtype=np.float32)
        tm = np.ascontiguousarray(tmask_u8, dtype=np.uint8)
        p_in = self._alloc(n * 3 * 4)
        p_tm = self._alloc(n)
        p_clean = self._alloc(n * 3)
        p_core = self._alloc(n)
        p_zone = self._alloc(n)
        try:
            self.mem.write(self.store, arr.tobytes(), p_in)
            self.mem.write(self.store, tm.tobytes(), p_tm)
            self.ex["deglow_full_green_v2"](
                self.store, p_in, h, w, p_tm,
                float(strength), float(zone_ratio), int(zone_expand),
                int(protect_px), int(chroma_keep), p_clean, p_core, p_zone)
            clean = np.frombuffer(bytes(self.mem.read(self.store, p_clean, p_clean + n * 3)),
                                 dtype=np.uint8).reshape(h, w, 3).copy()
            core = np.frombuffer(bytes(self.mem.read(self.store, p_core, p_core + n)),
                                 dtype=np.uint8).reshape(h, w).copy()
            zone = np.frombuffer(bytes(self.mem.read(self.store, p_zone, p_zone + n)),
                                 dtype=np.uint8).reshape(h, w).copy()
        finally:
            self._free(p_in, n * 3 * 4)
            self._free(p_tm, n)
            self._free(p_clean, n * 3)
            self._free(p_core, n)
            self._free(p_zone, n)
        return clean, core, zone

    def erase_text_glyphs(self, rgb_f32, h, w, tmask_u8, tmask2_u8=None,
                          strength: float = 1.0, zone_ratio: float = 0.6,
                          zone_expand: int = 0, protect_px: int = 0,
                          chroma_keep: int = 0, edge: int = 0,
                          direction_deg: float = -1.0, seed: int = 0,
                          edge_aware: int = 0, soft_expand: float = 0.0) -> tuple:
        """Single shared pipeline entry — run the FULL de-glow + mask-surgery +
        PatchMatch fill (browser + backend call this identically).

        rgb_f32   : H*W*3 float32 (0..255).
        tmask_u8  : H*W, >0 = known text (raw detect).
        tmask2_u8 : optional H*W second detect (on cleaned image) to union in.
        Returns (result H*W*3 u8, fill_mask H*W u8, clean H*W*3 u8, zone H*W u8).
        """
        n = h * w
        arr = np.ascontiguousarray(rgb_f32, dtype=np.float32)
        tm = np.ascontiguousarray(tmask_u8, dtype=np.uint8)
        if tmask2_u8 is None:
            tm2 = np.zeros((h, w), dtype=np.uint8)
        else:
            tm2 = np.ascontiguousarray(tmask2_u8, dtype=np.uint8)
        p_in = self._alloc(n * 3 * 4)
        p_tm = self._alloc(n)
        p_tm2 = self._alloc(n)
        p_result = self._alloc(n * 3)
        p_fill = self._alloc(n)
        p_clean = self._alloc(n * 3)
        p_zone = self._alloc(n)
        try:
            self.mem.write(self.store, arr.tobytes(), p_in)
            self.mem.write(self.store, tm.tobytes(), p_tm)
            self.mem.write(self.store, tm2.tobytes(), p_tm2)
            self.ex["erase_text_glyphs"](
                self.store, p_in, h, w, p_tm, p_tm2,
                float(strength), float(zone_ratio), int(zone_expand),
                int(protect_px), int(chroma_keep), int(edge),
                float(direction_deg), int(seed) & 0xFFFFFFFF,
                int(edge_aware), float(soft_expand),
                p_result, p_fill, p_clean, p_zone)
            result = np.frombuffer(bytes(self.mem.read(self.store, p_result, p_result + n * 3)),
                                   dtype=np.uint8).reshape(h, w, 3).copy()
            fill = np.frombuffer(bytes(self.mem.read(self.store, p_fill, p_fill + n)),
                                 dtype=np.uint8).reshape(h, w).copy()
            clean = np.frombuffer(bytes(self.mem.read(self.store, p_clean, p_clean + n * 3)),
                                  dtype=np.uint8).reshape(h, w, 3).copy()
            zone = np.frombuffer(bytes(self.mem.read(self.store, p_zone, p_zone + n)),
                                 dtype=np.uint8).reshape(h, w).copy()
        finally:
            self._free(p_in, n * 3 * 4)
            self._free(p_tm, n)
            self._free(p_tm2, n)
            self._free(p_result, n * 3)
            self._free(p_fill, n)
            self._free(p_clean, n * 3)
            self._free(p_zone, n)
        return result, fill, clean, zone


# Module-level singleton so the backend pays the load cost once.
_default = None


def get_core() -> TextCore:
    global _default
    if _default is None:
        _default = TextCore()
    return _default

File: shared/bindings/textcore.js
// Browser-side / Node binding for the shared WASM core (textcore.wasm).
//
// Single integration point the browser uses to call shared algorithms. Loads the
// SAME .wasm the Python backend loads (via wasmtime), so any operator added here
// is automatically available — and identical — on both ends.
//
// IMPORTANT: wasm memory may GROW during alloc / during a kernel call, which
// invalidates the `memory.buffer` reference. We re-fetch `ex.memory.buffer`
// right before every read/write. The browser worker must do the same.
const fs = require("fs");
const path = require("path");

const WASM_PATH = path.join(__dirname, "..", "build", "textcore.wasm");

let _instance = null;
let _ready = null;

function ready() {
  if (_ready) return _ready;
  _ready = WebAssembly.instantiate(fs.readFileSync(WASM_PATH), {}).then(
    ({ instance }) => {
      _instance = instance;
    }
  );
  return _ready;
}

async function distanceTransformEdt(maskU8, h, w) {
  await ready();
  const ex = _instance.exports;
  const n = h * w;
  const pMask = ex.alloc(n);
  const pOut = ex.alloc(n * 4);
  try {
    let m = ex.memory.buffer;
    new Uint8Array(m, pMask, n).set(maskU8);
    ex.distance_transform_edt(pMask, h, w, pOut);
    m = ex.memory.buffer; // re-fetch: kernel may have grown memory
    const out = new Float32Array(n);
    new Uint8Array(out.buffer).set(new Uint8Array(m, pOut, n * 4));
    ex.dealloc(pMask, n);
    ex.dealloc(pOut, n * 4);
    return out; // Float32Array length n
  } catch (e) {
    ex.dealloc(pMask, n);
    ex.dealloc(pOut, n * 4);
    throw e;
  }
}

async function rgbToGray(rgbF32, h, w) {
  await ready();
  const ex = _instance.exports;
  const n = h * w;
  const pIn = ex.alloc(n * 3 * 4);
  const pOut = ex.alloc(n);
  try {
    let m = ex.memory.buffer;
    new Float32Array(m, pIn, n * 3).set(rgbF32);
    ex.rgb_to_gray(pIn, h, w, pOut);
    m = ex.memory.buffer;
    const out = new Uint8Array(n);
    out.set(new Uint8Array(m, pOut, n));
    return out;
  } finally {
    ex.dealloc(pIn, n * 3 * 4);
    ex.dealloc(pOut, n);
  }
}

async function thresholdOtsu(u8, h, w) {
  await ready();
  const ex = _instance.exports;
  const n = h * w;
  const pIn = ex.alloc(n);
  const pOut = ex.alloc(n);
  try {
    let m = ex.memory.buffer;
    new Uint8Array(m, pIn, n).set(u8);
    const thr = ex.threshold_otsu(pIn, pOut, n);
    m = ex.memory.buffer;
    const bin = new Uint8Array(n);
    bin.set(new Uint8Array(m, pOut, n));
    return { thr, bin };
  } finally {
    ex.dealloc(pIn, n);
    ex.dealloc(pOut, n);
  }
}

async function morphology(maskU8, h, w, kernU8, kh, kw, op) {
  await ready();
  const ex = _instance.exports;
  const n = h * w;
  const nk = kh * kw;
  const opCode = op === "erode" ? 0 : 1;
  const pIn = ex.alloc(n);
  const pOut = ex.alloc(n);
  const pK = ex.alloc(nk);
  try {
    let m = ex.memory.buffer;
    new Uint8Array(m, pIn, n).set(maskU8);
    new Uint8Array(m, pK, nk).set(kernU8);
    ex.morphology(pIn, pOut, h, w, pK, kh, kw, opCode);
    m = ex.memory.buffer;
    const out = new Uint8Array(n);
    out.set(new Uint8Array(m, pOut, n));
    return out;
  } finally {
    ex.dealloc(pIn, n);
    ex.dealloc(pOut, n);
    ex.dealloc(pK, nk);
  }
}

async function connectedComponents(maskU8, h, w) {
  await ready();
  const ex = _instance.exports;
  const n = h * w;
  const pIn = ex.alloc(n);
  const pLabels = ex.alloc(n * 4);
  try {
    let m = ex.memory.buffer;
    new Uint8Array(m, pIn, n).set(maskU8);
    const ncomp = ex.connected_components(pIn, pLabels, h, w);
    m = ex.memory.buffer;
    const labels = new Int32Array(n);
    new Uint8Array(labels.buffer).set(new Uint8Array(m, pLabels, n * 4));
    const pStats = ex.alloc(ncomp * 5 * 4);
    try {
      ex.connected_components_stats(pLabels, pStats, h, w, ncomp);
      m = ex.memory.buffer;
      const raw = new Int32Array(ncomp * 5);
      new Uint8Array(raw.buffer).set(new Uint8Array(m, pStats, ncomp * 5 * 4));
      const stats = [];
      for (let i = 0; i < ncomp; i++) {
        const b = i * 5;
        stats.push({ left: raw[b], top: raw[b + 1], width: raw[b + 2], height: raw[b + 3], area: raw[b + 4] });
      }
      return { n: ncomp, labels, stats };
    } finally {
      ex.dealloc(pStats, ncomp * 5 * 4);
      ex.dealloc(pLabels, n * 4);
    }
  } finally {
    ex.dealloc(pIn, n);
  }
}

async function resizeGrayCubic(u8, h, w, h2, w2) {
  await ready();
  const ex = _instance.exports;
  const n = h * w;
  const n2 = h2 * w2;
  const pIn = ex.alloc(n);
  const pOut = ex.alloc(n2);
  try {
    let m = ex.memory.buffer;
    new Uint8Array(m, pIn, n).set(u8);
    ex.resize_gray_cubic(pIn, pOut, h, w, h2, w2);
    m = ex.memory.buffer;
    const out = new Uint8Array(n2);
    out.set(new Uint8Array(m, pOut, n2));
    return out;
  } finally {
    ex.dealloc(pIn, n);
    ex.dealloc(pOut, n2);
  }
}

async function resizeFloatLinear(f32, h, w, h2, w2) {
  await ready();
  const ex = _instance.exports;
  const n = h * w;
  const n2 = h2 * w2;
  const pIn = ex.alloc(n * 4);
  const pOut = ex.alloc(n2 * 4);
  try {
    let m = ex.memory.buffer;
    new Float32Array(m, pIn, n).set(f32);
    ex.resize_float_linear(pIn, pOut, h, w, h2, w2);
    m = ex.memory.buffer;
    const out = new Float32Array(n2);
    new Uint8Array(out.buffer).set(new Uint8Array(m, pOut, n2 * 4));
    return out;
  } finally {
    ex.dealloc(pIn, n * 4);
    ex.dealloc(pOut, n2 * 4);
  }
}

// THE shared PatchMatch fill. rgbF32: Float32Array H*W*3 (0..255, mutated to result).
async function patchmatchInpaint(rgbF32, h, w, maskU8, sampleU8, p, directionDeg, seed) {
  await ready();
  const ex = _instance.exports;
  const n = h * w;
  const pIn = ex.alloc(n * 3 * 4);
  const pOut = ex.alloc(n * 3 * 4);
  const pMask = ex.alloc(n);
  let pSample = 0;
  const hasSample = sampleU8 ? 1 : 0;
  if (sampleU8) pSample = ex.alloc(n);
  try {
    let m = ex.memory.buffer;
    new Float32Array(m, pIn, n * 3).set(rgbF32);
    new Uint8Array(m, pMask, n).set(maskU8);
    if (sampleU8) new Uint8Array(m, pSample, n).set(sampleU8);
    ex.patchmatch_inpaint(pIn, h, w, pMask, pSample, hasSample,
      p || 7, directionDeg == null ? -1.0 : directionDeg, (seed | 0) >>> 0, pOut);
    m = ex.memory.buffer;
    const out = new Float32Array(n * 3);
    new Uint8Array(out.buffer).set(new Uint8Array(m, pOut, n * 3 * 4));
    return out;
  } finally {
    ex.dealloc(pIn, n * 3 * 4);
    ex.dealloc(pOut, n * 3 * 4);
    ex.dealloc(pMask, n);
    if (pSample) ex.dealloc(pSample, n);
  }
}

// THE shared text-mask synthesis. textMaskU8: Uint8Array H*W (255=text).
// edge: >0 dilate / <0 erode / 0 identity (ellipse diameter = abs(edge)*2+1).
// limitU8: optional Uint8Array H*W (255=allowed fill). Returns [fillU8, sampleU8] 0/255.
async function synthesizeMasks(textMaskU8, h, w, edge, limitU8) {
  await ready();
  const ex = _instance.exports;
  const n = h * w;
  const pIn = ex.alloc(n);
  const pFill = ex.alloc(n);
  const pSmpl = ex.alloc(n);
  let pLim = 0;
  const hasLimit = limitU8 ? 1 : 0;
  if (limitU8) pLim = ex.alloc(n);
  try {
    let m = ex.memory.buffer;
    new Uint8Array(m, pIn, n).set(textMaskU8);
    if (limitU8) new Uint8Array(m, pLim, n).set(limitU8);
    ex.synthesize_masks(pIn, h, w, edge == null ? 1 : edge, pLim, hasLimit, pFill, pSmpl);
    m = ex.memory.buffer;
    const fill = new Uint8Array(n);
    const sample = new Uint8Array(n);
    fill.set(new Uint8Array(m, pFill, n));
    sample.set(new Uint8Array(m, pSmpl, n));
    return [fill, sample];
  } finally {
    ex.dealloc(pIn, n);
    ex.dealloc(pFill, n);
    ex.dealloc(pSmpl, n);
    if (pLim) ex.dealloc(pLim, n);
  }
}

// THE shared de-glow (full green v2). rgbF32: Float32Array H*W*3 (0..255).
// tmaskU8: Uint8Array H*W (255=text). Returns [cleanU8 H*W*3, coreU8 H*W] (0..255/0|255).
async function deglowFullGreenV2(rgbF32, h, w, tmaskU8, strength, zoneRatio, zoneExpand, protectPx, chromaKeep) {
  await ready();
  const ex = _instance.exports;
  const n = h * w;
  const pIn = ex.alloc(n * 3 * 4);
  const pTm = ex.alloc(n);
  const pClean = ex.alloc(n * 3);
  const pCore = ex.alloc(n);
  const pZone = ex.alloc(n);
  try {
    let m = ex.memory.buffer;
    new Float32Array(m, pIn, n * 3).set(rgbF32);
    new Uint8Array(m, pTm, n).set(tmaskU8);
    ex.deglow_full_green_v2(pIn, h, w, pTm,
      strength == null ? 1.0 : strength,
      zoneRatio == null ? 0.6 : zoneRatio,
      zoneExpand == null ? 0 : zoneExpand,
      protectPx == null ? 0 : protectPx,
      chromaKeep == null ? 0 : chromaKeep,
      pClean, pCore, pZone);
    m = ex.memory.buffer;
    const clean = new Uint8Array(n * 3);
    const core = new Uint8Array(n);
    const zone = new Uint8Array(n);
    clean.set(new Uint8Array(m, pClean, n * 3));
    core.set(new Uint8Array(m, pCore, n));
    zone.set(new Uint8Array(m, pZone, n));
    return [clean, core, zone];
  } finally {
    ex.dealloc(pIn, n * 3 * 4);
    ex.dealloc(pTm, n);
    ex.dealloc(pClean, n * 3);
    ex.dealloc(pCore, n);
    ex.dealloc(pZone, n);
  }
}

// THE single shared pipeline entry. Runs the FULL de-glow + mask-surgery +
// PatchMatch fill (browser + backend call this identically).
// rgbF32: Float32Array H*W*3 (0..255). tmaskU8: Uint8Array H*W (255=text).
// tmask2U8: optional Uint8Array H*W second detect to union in.
// Returns [resultU8 H*W*3, fillU8 H*W, cleanU8 H*W*3, zoneU8 H*W].
async function eraseTextGlyphs(rgbF32, h, w, tmaskU8, tmask2U8,
  strength, zoneRatio, zoneExpand, protectPx, chromaKeep, edge, directionDeg, seed,
  edgeAware, softExpand) {
  await ready();
  const ex = _instance.exports;
  const n = h * w;
  const pIn = ex.alloc(n * 3 * 4);
  const pTm = ex.alloc(n);
  const pTm2 = ex.alloc(n);
  const pResult = ex.alloc(n * 3);
  const pFill = ex.alloc(n);
  const pClean = ex.alloc(n * 3);
  const pZone = ex.alloc(n);
  try {
    let m = ex.memory.buffer;
    new Float32Array(m, pIn, n * 3).set(rgbF32);
    new Uint8Array(m, pTm, n).set(tmaskU8);
    if (tmask2U8) new Uint8Array(m, pTm2, n).set(tmask2U8);
    ex.erase_text_glyphs(pIn, h, w, pTm, pTm2,
      strength == null ? 1.0 : strength,
      zoneRatio == null ? 0.6 : zoneRatio,
      zoneExpand == null ? 0 : zoneExpand,
      protectPx == null ? 0 : protectPx,
      chromaKeep == null ? 0 : chromaKeep,
      edge == null ? 0 : edge,
      directionDeg == null ? -1.0 : directionDeg,
      (seed == null ? 0 : seed) >>> 0,
      edgeAware == null ? 0 : edgeAware,
      softExpand == null ? 0.0 : softExpand,
      pResult, pFill, pClean, pZone);
    m = ex.memory.buffer;
    const result = new Uint8Array(n * 3);
    const fill = new Uint8Array(n);
    const clean = new Uint8Array(n * 3);
    const zone = new Uint8Array(n);
    result.set(new Uint8Array(m, pResult, n * 3));
    fill.set(new Uint8Array(m, pFill, n));
    clean.set(new Uint8Array(m, pClean, n * 3));
    zone.set(new Uint8Array(m, pZone, n));
    return [result, fill, clean, zone];
  } finally {
    ex.dealloc(pIn, n * 3 * 4);
    ex.dealloc(pTm, n);
    ex.dealloc(pTm2, n);
    ex.dealloc(pResult, n * 3);
    ex.dealloc(pFill, n);
    ex.dealloc(pClean, n * 3);
    ex.dealloc(pZone, n);
  }
}

module.exports = {
  ready,
  distanceTransformEdt,
  rgbToGray,
  thresholdOtsu,
  morphology,
  connectedComponents,
  resizeGrayCubic,
  resizeFloatLinear,
  patchmatchInpaint,
  synthesizeMasks,
  deglowFullGreenV2,
  eraseTextGlyphs,
  WASM_PATH,
};
